src/v/raft/consensus.cc

```diff
@@ -1988,11 +1988,55 @@ consensus::do_append_entries(append_entries_request&& r) {
         co_return reply;
     }
 
-    // special case heartbeat case
+    model::offset batch_prev_log_index = request_metadata.prev_log_index;
+    if (batch_prev_log_index < last_log_offset) {
+        // The append point is before the end of our log. We need to skip
+        // over batches that we already have (they will have the matching
+        // term) to find the true truncation point. This is important for the
+        // case when we already have _all_ batches locally (possible if e.g.
+        // the request was delayed/duplicated). In this case we don't want to
+        // truncate, otherwise we might lose already committed data.
+
+        struct find_mismatch_consumer {
+            ss::future<ss::stop_iteration>
+            operator()(const model::record_batch& b) {
+                model::offset last_offset = last_matched
+                                            + model::offset(
+                                              b.header().last_offset_delta + 1);
+                if (parent.get_term(last_offset) != b.term()) {
+                    co_return ss::stop_iteration::yes;
+                }
+                last_matched = last_offset;
+                co_return ss::stop_iteration::no;
+            }
+
+            model::offset end_of_stream() { return last_matched; }
+
+            consensus& parent;
+            model::offset last_matched;
+        };
+
+        model::offset last_matched = co_await r.batches().peek_each_ref(
+          find_mismatch_consumer{
+            .parent = *this, .last_matched = batch_prev_log_index},
+          model::no_timeout); // no_timeout as the batches are already in memory
+        if (last_matched != batch_prev_log_index) {
+            vlog(
+              _ctxlog.info,
+              "skipped matching records in append_entries batch from {} to {}, "
+              "current state: {}",
+              batch_prev_log_index,
+              last_matched,
+              meta());
+            batch_prev_log_index = last_matched;
+        }
+    }
+
+    // special case for heartbeats and batches without new records.
     // we need to handle it early (before executing truncation)
     // as timeouts are asynchronous to append calls and can have stall data
     if (r.batches().is_end_of_stream()) {
-        if (request_metadata.prev_log_index < last_log_offset) {
+        if (batch_prev_log_index < last_log_offset) {
```

Comment:
nit: I was wondering if there is a better name for `batch_prev_log_index`, like `adjusted_prev_log_index` or some such.. I feel like `batch_prev_log_index` doesn't convey full intent.